contracts/compliance/modular/modules/InvestorCountryCapModule.sol

```diff
@@ -0,0 +1,332 @@
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ *     You should have received a copy of the GNU General Public License
+ *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ *     This specific smart contract is also licensed under the Creative Commons
+ *     Attribution-NonCommercial 4.0 International License (CC-BY-NC-4.0),
+ *     which prohibits commercial use. For commercial inquiries, please contact
+ *     Tokeny sàrl for licensing options.
+ */
+
+pragma solidity 0.8.27;
+
+import "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+import "./AbstractModuleUpgradeable.sol";
+import "../IModularCompliance.sol";
+import "../../../token/IToken.sol";
+import "../../../roles/AgentRole.sol";
+
+event CountryCapSet(uint16 indexed country, uint256 cap);
+event BypassedIdentityAdded(address indexed identity);
+event BypassedIdentityRemoved(address indexed identity);
+
+error ExpectedPause();
+error IdentityNotBypassed(address identity);
+error CapLowerThanCurrent(uint16 country, uint256 cap, uint256 currentCap);
+error WalletCountLimitReached(address identity, uint256 maxWallets);
+
+uint256 constant MAX_WALLET_PER_IDENTITY = 20;
+
+
+contract InvestorCountryCapModule is AbstractModuleUpgradeable {
+    using EnumerableSet for EnumerableSet.AddressSet;
+    using EnumerableSet for EnumerableSet.UintSet;
+
+    struct CountryParams {
+        bool capped;
+        uint256 cap;
+        uint256 count;
+        mapping(address identity => bool counted) identities;
+    }
+
+    EnumerableSet.UintSet internal _countries;
+    mapping(address identity => bool bypassed) internal _bypassedIdentities;
+
+    mapping(address compliance => mapping(uint16 country => CountryParams params)) internal _countryParams;
+    mapping(address compliance => mapping(address identity => EnumerableSet.AddressSet wallets)) internal _identityToWallets;
+
+    /// @notice Used only during batchInitialize / canComplianceBind
+    mapping(address token => uint256 supply) public calculatedSupply;
+    
+    
+    /// @dev initializes the contract and sets the initial state.
+    /// @notice This function should only be called once during the contract deployment, and after (optionally) batchInitialize.
+    function initialize() external initializer {
+        __AbstractModule_init();
+    }
+
+    /// @dev Initialize the module for a compliance and a list of holders
+    /// @param _compliance Address of the compliance.
+    /// @param _holders Addresses of the holders already holding tokens (addresses should be unique - no control is done on that).
+    function batchInitialize(address _compliance, address[] memory _holders) external onlyOwner {
+        // TODO calculate gas cost and revert if _holders.length is too high
```

Comment:
what about this todo? did you have time to check? normally the gas reporter should give a good estimate